sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/rules/RuleExecutor.scala

```diff
@@ -47,6 +47,9 @@ abstract class RuleExecutor[TreeType <: TreeNode[_]] extends Logging {
    */
   abstract class Strategy { def maxIterations: Int }
 
+  /** A strategy that is idempotent, i.e. additional runs should not change query plans */
+  case object Idempotent extends Strategy { val maxIterations = 1 }
+
   /** A strategy that only runs once. */
   case object Once extends Strategy { val maxIterations = 1 }
```

Comment:
now we have both `Once` and `Idempotent`, when to pick which?

Reply:
I would suggest to use `Once` in the test suites only. 
 
In the unit tests of optimizer rules, we have many batches that are using `Once`. The expected result of the optimized plan is based on the Once policy. These batches are not expected to be idempotent.

Reply:
To be clear, `Once` means a batch should be run only once, but does not guarantee anything for multiple runs. On the other hand, `Idempotent` means that a batch should be run once, but multiple runs should yield the same result of a single run.

Reply:
Basically if a batch is marked as `Idempotent`, the idempotence contract is enforced by running an additional checker.

Reply:
After we make sure all rules are idempotent, `Once` would be for test only and the names would all seem weird and confusing.
One way is to use `Once` for the semantics of idempotent, and add a test-only `OnceUnchecked` for all the catalyst rule tests (I don't suppose we should use this anywhere else).
Another option is to add another ENV variable, similar to `SPARK_TESTING` (sth. like `SPARK_RULE_TESTING`) for the `RuleExecutor` to skip the check.

Reply:
See `RuleExecutorSuite`, `Once` *was* supposed to run only once (and does not have to necessarily pass the sanity check).

Reply:
I know what you are saying, @yeshengm. All I meant was all we need is `Idempotent` and `FixedPoint` except for the occasions that @gatorsmile worried about in our catalyst rule tests where really `FixedPoint` should be used but it was mistakenly put as `Once`.

Now I remember last time I worked the experimental branch, there wasn't a whole lot of such places, so let's just change those false `Once` to `FixedPoint` in our tests, and we'll need `Once` only (which is checked under testing mode).

Reply:
@maryannxue We can definitely change that. But that means we also have to change tests in `RuleExecutorSuite`. So you are suggesting that we only have `Once` with idempotence semantics & `FixedPoint` right?

Reply:
Yep.